internal/dataplane/kongstate/config_fetcher.go

```diff
@@ -0,0 +1,144 @@
+package kongstate
+
+import (
+	"context"
+
+	"github.com/kong/deck/dump"
+	"github.com/kong/deck/utils"
+	"github.com/kong/go-kong/kong"
+)
+
+type KongLastGoodConfigFetcher interface {
+	// GetKongRawState fetches the configuration loaded in a Kong node.
+	GetKongRawState(ctx context.Context, client *kong.Client) (*utils.KongRawState, error)
+	// GetKongStatus fetches the status of a Kong node.
+	GetKongStatus(ctx context.Context, client *kong.Client) (*kong.Status, error)
+}
```

Comment:
Idea: given we already have created some kind of a separate component responsible for the last valid config story (`KongLastGoodConfigFetcher`), maybe we could move the whole logic responsible for fetching the config from all gateway clients to this. This would allow easier unit testing and we could avoid ever-growing and noisy `KongClient` code and unit tests (they would be simplified to assertions that ensure the new component is called when needed). In the new component's tests on the other hand we could focus on ensuring the logic of looping over all gateways is correct and so on.

Its interface could be: 
```go
type LastGoodConfigFetcher interface {
	// TryFetchingValidConfigFromGateways tries to fetch a valid configuration from all gateways and persists it if found.
	TryFetchingValidConfigFromGateways(ctx context.Context) error

	// LastValidConfig returns the last valid config and true if there's one available. Otherwise, second return value is false.
        LastValidConfig(ctx context.Context) (*KongState, bool)

	// StoreLastValidConfig stores a given configuration as the last valid config. Should be used when the configuration was successfully accepted by a gateway.
	StoreLastValidConfig(s *KongState)
}
```
It could accept a `GatewayClients() []*adminapi.Client` interface as a dependency (that's implemented by `ClientsManager`) and use it to list clients in `TryFetchingValidConfigFromGateways` method. 

Note: please tell me if I'm too picky -- I'm trying to think about how we can make this as encapsulated as possible to make future modifications easier.

Reply:
It looks like a good improvement to me 👍 I've implemented it here: ba25db208a40dec52bcfe481d9b10136abc3e951

Reply:
Looks good to me! Thank you!